File: C++/dao/ModeloDao.cpp
#include "ModeloDao.h"

// Construtor para abrir o banco
ModeloDao::ModeloDao(const std::string& dbName) {
    if (sqlite3_open(dbName.c_str(), &db)) {
        std::cerr << "Não foi possível abrir o banco de dados: " << sqlite3_errmsg(db) << std::endl;
    }
}

// Destrutor para fechar o banco
ModeloDao::~ModeloDao() {
    sqlite3_close(db);
}

// Buscar uma modelo pelo ID
Modelo ModeloDao::findById(int idModelo) {
    std::string sql = "SELECT idModelo, idMarca, nome FROM Modelo WHERE idModelo = ?";
    sqlite3_stmt* stmt;

    if (sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt, 0) != SQLITE_OK) {
        throw std::runtime_error("Erro ao preparar consulta: " + std::string(sqlite3_errmsg(db)));
    }

    sqlite3_bind_int(stmt, 1, idModelo);

    Modelo modelo(0, 0, ""); // Objeto vazio para armazenar os dados
    if (sqlite3_step(stmt) == SQLITE_ROW) {
        modelo.setIdModelo(sqlite3_column_int(stmt, 0));
        modelo.setIdMarca(sqlite3_column_int(stmt, 1));
        modelo.setNome(reinterpret_cast<const char*>(sqlite3_column_text(stmt, 2)));
    } else {
        std::cerr << "Nenhuma modelo encontrada com o ID: " << idModelo << std::endl;
    }

    sqlite3_finalize(stmt);
    return modelo;
}

// Buscar todas as modelos
std::vector<Modelo> ModeloDao::findAll() {
    std::string sql = "SELECT idModelo, idMarca, nome FROM Modelo";
    sqlite3_stmt* stmt;

    if (sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt, 0) != SQLITE_OK) {
        throw std::runtime_error("Erro ao preparar consulta: " + std::string(sqlite3_errmsg(db)));
    }

    std::vector<Modelo> modelos;
    while (sqlite3_step(stmt) == SQLITE_ROW) {
        Modelo modelo(
            sqlite3_column_int(stmt, 0),
            sqlite3_column_int(stmt, 1),
            reinterpret_cast<const char*>(sqlite3_column_text(stmt, 2))
        );
        modelos.push_back(modelo);
    }

    sqlite3_finalize(stmt);
    return modelos;
}

// Inserir um novO modelo
void ModeloDao::insert(const Modelo& modelo) {
    std::string sql = "INSERT INTO Modelo (idMarca, nome) VALUES (?, ?)";
    sqlite3_stmt* stmt;

    if (sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt, 0) != SQLITE_OK) {
        throw std::runtime_error("Erro ao preparar consulta: " + std::string(sqlite3_errmsg(db)));
    }

    sqlite3_bind_int(stmt, 1, modelo.getIdMarca());
    sqlite3_bind_text(stmt, 2, modelo.getNome().c_str(), -1, SQLITE_STATIC);

    if (sqlite3_step(stmt) != SQLITE_DONE) {
        std::cerr << "Erro ao inserir a modelo: " << sqlite3_errmsg(db) << std::endl;
    }

    sqlite3_finalize(stmt);
}

// Atualizar uma modelo existente
void ModeloDao::update(const Modelo& modelo) {
    std::string sql = "UPDATE Modelo SET idMarca = ?, nome = ? WHERE idModelo = ?";
    sqlite3_stmt* stmt;

    if (sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt, 0) != SQLITE_OK) {
        throw std::runtime_error("Erro ao preparar consulta: " + std::string(sqlite3_errmsg(db)));
    }

    sqlite3_bind_int(stmt, 1, modelo.getIdMarca());
    sqlite3_bind_text(stmt, 2, modelo.getNome().c_str(), -1, SQLITE_STATIC);
    sqlite3_bind_int(stmt, 3, modelo.getIdModelo());

    if (sqlite3_step(stmt) != SQLITE_DONE) {
        std::cerr << "Erro ao atualizar a modelo: " << sqlite3_errmsg(db) << std::endl;
    }

    sqlite3_finalize(stmt);
}

// Deletar uma modelo pelo ID
void ModeloDao::deleteById(int idModelo) {
    std::string sql = "DELETE FROM Modelo WHERE idModelo = ?";
    sqlite3_stmt* stmt;

    if (sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt, 0) != SQLITE_OK) {
        throw std::runtime_error("Erro ao preparar consulta: " + std::string(sqlite3_errmsg(db)));
    }

    sqlite3_bind_int(stmt, 1, idModelo);

    if (sqlite3_step(stmt) != SQLITE_DONE) {
        std::cerr << "Erro ao deletar a modelo: " << sqlite3_errmsg(db) << std::endl;
    }

    sqlite3_finalize(stmt);
}

File: JAVA/src/main/java/com/inf_112/trabalho_final/dao/ModeloDao.java
package com.inf_112.trabalho_final.dao;

import com.inf_112.trabalho_final.model.Modelo;
import java.util.List;
import java.util.Optional;

import org.springframework.jdbc.core.simple.JdbcClient;
import org.springframework.stereotype.Repository;

@Repository
public class ModeloDao {
        private final JdbcClient jdbcClient;

        public ModeloDao(JdbcClient jdbcClient) {
                this.jdbcClient = jdbcClient;
        }

        // Buscar uma Modelo pelo ID
        public Modelo findById(int idModelo) {
                Optional<Modelo> optionalModelo = jdbcClient
                                .sql("SELECT * FROM Modelo WHERE idModelo = :idModelo")
                                .param("idModelo", idModelo)
                                .query(Modelo.class)
                                .optional();

                if (optionalModelo.isPresent()) {
                        return optionalModelo.get();
                }

                return null;
        }

        // Buscar todas as Modeloes
        public List<Modelo> findAllByMarca(int idMarca) {
                return jdbcClient
                                .sql("SELECT * FROM Modelo WHERE idMarca = :idMarca")
                                .param("idMarca", idMarca)
                                .query(Modelo.class)
                                .list();
        }

        // Buscar todas as Modeloes
        public List<Modelo> findAll() {
                return jdbcClient
                                .sql("SELECT * FROM Modelo")
                                .query(Modelo.class)
                                .list();
        }

        // Inserir uma nova Modelo
        public void insert(Modelo modelo) {
                jdbcClient
                                .sql("INSERT INTO Modelo (nome, idMarca) VALUES (:nome, :idMarca)")
                                .param("nome", modelo.getNome())
                                .param("idMarca", modelo.getIdMarca())
                                .update();
        }

        // Atualizar uma Modelo existente
        public void update(Modelo Modelo) {
                jdbcClient
                                .sql("UPDATE Modelo SET nome = :nome, idMarca = :idMarca WHERE idModelo = :idModelo")
                                .param("nome", Modelo.getNome())
                                .param("idMarca", Modelo.getIdMarca())
                                .param("idModelo", Modelo.getIdModelo())
                                .update();
        }

        // Deletar uma Modelo pelo ID
        public void delete(int idModelo) {
                jdbcClient
                                .sql("DELETE FROM Modelo WHERE idModelo = :idModelo")
                                .param("idModelo", idModelo)
                                .update();
        }
}